general/testing/README.md

```diff
@@ -0,0 +1,183 @@
+
+#### Assert (ie. `expect()`) one thing per `it()` block
+
+Instead of:
+
+```javascript
+describe('prepareConfig()', => {
+  it('generates the right config', () => {
+    const config = prepareConfig({preset: 'default'});
+
+    expect(config).not.toBe(null);
+    expect(Object.isFrozen(config)).toBe(true);
+    expect(config.plugins).toEqual([]);
+
+    expect(() => prepareConfig({preset: 'standard'}))
+      .toThrow(/Unknown preset/);
+  });
+});
+```
+
+write:
+
+```javascript
+describe('prepareConfig()', => {
+  let config;
+
+  beforeEach(() => {
+    config = prepareConfig({preset: 'default'});
+  });
+
+  it('returns a non-null configuration', () => {
+    expect(config).not.toBe(null);
+  });
+
+  it('returns a frozen object', () => {
+    expect(Object.isFrozen(config)).toBe(true);
+  });
+
+  it('includes plugin configuration', () => {
+    expect(config.plugins).toEqual([]);
+  });
+
+  it('rejects unknown preset names', () => {
+    expect(() => prepareConfig({preset: 'standard'}))
+      .toThrow(/Unknown preset/);
+  });
+});
+```
+
+By creating small, targeted `it()` blocks:
+
+-   A failure early on in the `it()` block does not preclude the other assertions from being evaluated.
+-   These assertions more readable if you look at it in isolation (because each one is simpler).
+-   They are also more intention-revealing.
+-   The test output now serves as better documentation of the expected behavior, because we see:
+
+    ```
+      prepareConfig()
+        ✓ returns a non-null configuration (54ms)
+        ✓ returns a frozen object (3ms)
+        ✓ includes plugin configuration (10ms)
+        ✓ rejects unknown preset names (4ms)
+    ```
+
+    instead of just:
+
+    ```
+      prepareConfig()
+        ✓ generates the right config (103ms)
+    ```
+
+#### Don't dynamically create `describe()` and `it()` blocks
+
+Instead of:
+
+```javascript
+['left', 'right', 'center'].forEach(position => {
+	describe(`positioning controls at the ${position}`, () => {
+		it('has a matching position property', () => {
+			expect(getControl(position).position).toBe(position);
+		});
+	});
+});
+```
+
+write:
+
+```javascript
+describe('positioning controls to the left', () => {
+	it('has a position property of "left"', () => {
+		expect(getControl('left').position).toBe('left');
+	});
+});
+
+describe('positioning controls to the right', () => {
+	it('has a position property of "right"', () => {
+		expect(getControl('right').position).toBe('right');
+	});
+});
+
+describe('positioning controls in the centre', () => {
+	it('has a position property of "center"', () => {
+		expect(getControl('center').position).toBe('center');
+	});
+});
+```
+
+This (contrived) example is one where your intuition may be to "DRY" up the code, but consider the position of the person who reads a report of a "test failure on line 123" and now has to figure out which iteration of the loop. It's better for them if we spell out each example alone.
```

Comment:
nit: maybe add a link to the DRY section?

Reply:
Something like this?

https://github.com/liferay/liferay-frontend-guidelines/pull/63/commits/1f13ba3321ed73b9bf1e2e1bb80545d0fd273118

I'll merge this with that, and if I misunderstood what you meant please feel free to edit.